Add missing CONTRIBUTING.md file

Hi,

I want to propose the idea of having `CONTRIBUTING.md` file so that developers (especially new Haskell developers) can get useful guide on how to build, run and test the project on their local machines.

#### Helpful pointers that could be added:
- Whether to use cabal or stack?
- If cabal, then do we use v1 commands or are `cabal new-*` commands recommended ?
- If stack, then what should be the value of resolver?
- How to run a particular test instead of all tests?
- How to debug, make changes and see the changes in action?

I understand some of these things are quite basic and to some extent repetitive but adding these pointers are one way of learning about Haskell projects for beginners. This can definitely bring more developers to trying out new projects and bring obscure bugs to project maintainers notice.

I am currently facing problems in trying to setup the project locally and hence I have raised this issue. I tried to build the project using `cabal new-build` which produced the following error:
```
$ cabal new-build
Resolving dependencies...
cabal: Could not resolve dependencies:
[__0] trying: ormolu-0.0.1.0 (user goal)
[__1] trying: gitrev-1.3.1 (dependency of ormolu)
[__2] next goal: template-haskell (dependency of gitrev)
[__2] rejecting: template-haskell-2.11.1.0/installed-2.1... (conflict: ormolu
=> ghc-boot-th>=8.4.3, template-haskell =>
ghc-boot-th==8.0.2/installed-8.0...)
[__2] rejecting: template-haskell-2.14.0.0, template-haskell-2.13.0.0,
template-haskell-2.12.0.0, template-haskell-2.11.1.0,
template-haskell-2.11.0.0, template-haskell-2.10.0.0,
template-haskell-2.9.0.0, template-haskell-2.8.0.0, template-haskell-2.7.0.0,
template-haskell-2.6.0.0, template-haskell-2.5.0.0, template-haskell-2.4.0.1,
template-haskell-2.4.0.0, template-haskell-2.3.0.1, template-haskell-2.3.0.0,
template-haskell-2.2.0.0 (constraint from non-upgradeable package requires
installed instance)
[__2] fail (backjumping, conflict set: gitrev, ormolu, template-haskell)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: template-haskell, ormolu, gitrev
```
I tried to remove `~./ghc` and `~/.cabal` directories, guessing that it could be due to conflict in the dependencies and re-ran `cabal new-build` but it failed.

Later I created `stack.yaml` and included `resolver: lts-13.19` but ran into different errors this time. Here's the short snippet of error-log (complete log is in this [link](https://gist.github.com/abiduzz420/7f60529e381e7faf607aa4c4a84f4804)) :
```
$ stack build
Building all executables for `ormolu' once. After a successful build of all of them, only specified executables will be rebuilt.
ormolu-0.0.1.0: build (lib + exe)
Preprocessing library for ormolu-0.0.1.0..
Building library for ormolu-0.0.1.0..
[11 of 24] Compiling Ormolu.Printer.Meat.Type ( src/Ormolu/Printer/Meat/Type.hs, .stack-work/dist/x86_64-linux/Cabal-2.4.0.1/build/Ormolu/Printer/Meat/Type.o )

/home/abiduzair/Documents/uzair/ormolu/src/Ormolu/Printer/Meat/Type.hs:36:3: error:
    Not in scope: data constructor ‘HsAppsTy’
    Perhaps you meant ‘HsAppTy’ (imported from GHC)
   |
36 |   HsAppsTy apps ->
   |   ^^^^^^^^

/home/abiduzair/Documents/uzair/ormolu/src/Ormolu/Printer/Meat/Type.hs:53:3: error:
    Not in scope: data constructor ‘HsPArrTy’
    Perhaps you meant ‘HsParTy’ (imported from GHC)
   |
53 |   HsPArrTy t -> located t (bracketsPar . p_hsType)
   |   ^^^^^^^^
...
...
...
--  While building package ormolu-0.0.1.0 using:
      /home/abiduzair/.stack/setup-exe-cache/x86_64-linux/Cabal-simple_mPHDZzAJ_2.4.0.1_ghc-8.6.4 --builddir=.stack-work/dist/x86_64-linux/Cabal-2.4.0.1 build lib:ormolu exe:ormolu --ghc-options " -ddump-hi -ddump-to-file -fdiagnostics-color=always"
    Process exited with code: ExitFailure 1
```
I would like to know what are your thoughts on adding `CONTRIBUTING.md` file and how I could resolve the project setup issues mentioned above. I am happy to raise a PR for the same, if I can successfully run the project locally. Thank you :)

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -7,6 +7,11 @@ jobs:
     steps:
       - checkout
       - run:
+          name: Add ca-certificates
+          command: |
+            apk --no-progress update
+            apk --no-progress add ca-certificates
+      - run:
           name: Setup Cachix
           command: |
             nix-env -f nix/nixpkgs/ -iA cachix
@@ -22,6 +27,12 @@ jobs:
       - run:
           name: Build the package (includes running the tests)
           command: nix-build
+      - run:
+          name: Generate Haddocks
+          command: nix-build --attr doc
+      - store_artifacts:
+          path: result-doc/share/doc
+          destination: haddock
 
 workflows:
   version: 2

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -1,0 +1,70 @@
+# Contributing
+
+Issues (bugs, feature requests or otherwise feedback) may be reported in
+[the GitHub issue tracker for this
+project](https://github.com/tweag/ormolu/issues). Pull requests are also
+welcome.
+
+When contributing to this repository, please first discuss the change you
+wish to make via an issue, unless it's entirely trivial (typo fixes, etc.).
+If there is already an issue that describes the change you have in mind,
+comment on it indicating that you're going to work on that. This way we can
+avoid the situation when several people work on the same thing.
+
+Please make sure that all non-trivial changes are described in commit
+messages and PR descriptions.
+
+## What to hack on?
+
+Right now there are two options for people who want to contribute:
+
+* [Implementing rendering of AST][cover-ast]. This is the main focus right
+  now because we want to have a MVP which can render all syntactical
+  constructions found in Haskell source code. Once that is achieved we'll
+  polish the tool iteratively.
+* [Fixing bugs][bugs]. If this seems more interesting than implementing
+  rendering of AST, you are welcome to do this as well.
+
+## Implementing rendering of AST
+
+The `Ormolu.Printer.Combinators` module provides a DSL for rendering of GHC
+AST. You'll probably only need this one module for writing new rendering
+functions. The module documents how to use the printing combinators it
+provides. Consult the Haddocks to learn more about them.
+
+Create new modules corresponding to the things you want to render under
+`Ormolu.Printer.Meat`. For example, there are `Ormolu.Printer.Meat.Type` and
+`Ormolu.Printer.Meat.Declaration.Data`.
+
+Concrete rendering functions get their names by appending `p_` to the name
+of the type which describes a particular part of AST. For example, the
+function for rendering types is correspondingly:
+
+```haskell
+p_hsType :: HsType GhcPs -> R ()
+```
+
+In general rendering functions like this take 1 or more arguments and
+produce `R ()` which is a rendering action.
+
+### Testing
+
+Testing has been taken good care of and now it amounts to just adding
+examples under `data/examples`. Each example is a pair of files:
+`<example-name>.hs` for input and `<example-name>-out.hs` for corresponding
+expected output.
+
+Testing is performed as following:
+
+* Given snippet of source code is parsed and pretty-printed.
+* The result of printing is parsed back again and the AST is compared to the
+  AST obtained from the original file. They should match.
+* The output of printer is checked against the expected output.
+* Idempotency property is verified: formatting already formatted code
+  results in exactly the same output.
+
+Examples can be organized in sub-directories, see the existing ones for
+inspiration.
+
+[cover-ast]: https://github.com/tweag/ormolu/issues?q=is%3Aissue+is%3Aopen+label%3Acovering-ghc-ast
+[bugs]: https://github.com/tweag/ormolu/issues?q=is%3Aissue+is%3Aopen+label%3Abug

--- a/README.md
+++ b/README.md
@@ -76,72 +76,15 @@ debug: False
 $ ormolu -c ormolu.yaml Module.hs
 ```
 
-## Contribution
+## Contributing
 
-Issues (bugs, feature requests or otherwise feedback) may be reported in
-[the GitHub issue tracker for this
-project](https://github.com/tweag/ormolu/issues). Pull requests are also
-welcome.
-
-### What to hack on?
-
-Right now there are two options for people who want to contribute:
-
-* [Implementing rendering of AST][cover-ast]. This is the main focus right
-  now because we want to have a MVP which can render all syntactical
-  constructions found in Haskell source code. Once that is achieved we'll
-  polish the tool iteratively.
-* [Fixing bugs][bugs]. If this seems more interesting than implementing
-  rendering of AST, you are welcome to do this as well.
-
-### Implementing rendering of AST
-
-The `Ormolu.Printer.Combinators` module provides a DSL for rendering of GHC
-AST. You'll probably only need this one module for writing new rendering
-functions. The module documents how to use the printing combinators it
-provides. Consult the Haddocks to learn more about them.
-
-Create new modules corresponding to the things you want to render under
-`Ormolu.Printer.Meat`. For example, there are `Ormolu.Printer.Meat.Type` and
-`Ormolu.Printer.Meat.Declaration.Data`.
-
-Concrete rendering functions get their names by appending `p_` to the name
-of the type which describes a particular part of AST. For example, the
-function for rendering types is correspondingly:
-
-```haskell
-p_hsType :: HsType GhcPs -> R ()
-```
-
-In general rendering functions like this take 1 or more arguments and
-produce `R ()` which is a rendering action.
-
-### Testing
-
-Testing has been taken good care of and now it amounts to just adding
-examples under `data/examples`. Each example is a pair of files:
-`<example-name>.hs` for input and `<example-name>-out.hs` for corresponding
-expected output.
-
-Testing is performed as following:
-
-* Given snippet of source code is parsed and pretty-printed.
-* The result of printing is parsed back again and the AST is compared to the
-  AST obtained from the original file. They should match.
-* The output of printer is checked against the expected output.
-* Idempotency property is verified: formatting already formatted code
-  results in exactly the same output.
-
-Examples can be organized in sub-directories, see the existing ones for
-inspiration.
+See [CONTRIBUTING.md](./CONTRIBUTING.md).
 
 ## License
 
-See [LICENSE](./LICENSE.md).
+See [LICENSE.md](./LICENSE.md).
 
 Copyright © 2018–2019 Tweag I/O
 
 [haskell-src-exts]: https://hackage.haskell.org/package/haskell-src-exts
 [hindent]: https://hackage.haskell.org/package/hindent
-[cover-ast]: https://github.com/tweag/ormolu/issues?q=is%3Aissue+is%3Aopen+label%3Acovering-ghc-ast
-[bugs]: https://github.com/tweag/ormolu/issues?q=is%3Aissue+is%3Aopen+label%3Abug

